T vector3d api 56

brb1031 opened the pull request:
The goal of this PR is to get feedback on the API for Vector3d.

The majority of this branch is new testing code for the Vector3d class.  Many tweaks were made to the implementation of Vector3d to improve numerical stability and robustness.  Any obviously unnecessary calculations were removed.

There are a number of functions that were implemented because they might be useful,   Looking at setFromV, setXYZ, the constructor, and .copy(), it seems like there must be some way to make the interface more streamlined.

brb1031 committed:
Add vector math tests, standardize function names

Performs tests on randomly generated vectors for correctness of known
relations. Mutators are renamed to contain "Me", ie "rotateMe",
"incrementMe", etc. Functions are reordered such that "setters" are
adjacent to "getters".  Vector3d.js and its tests are linted.

API:

Added a short block for aliasing the functions with the old names.
Added compatibility function wrapper for rotateAbout (just to keep
things running and testable).  This calls rotateMe, which takes an
object with rotation details: pivotPoint (location in space about
which to rotate), axis: the axis of rotation (it can be thought of as
passing through the pivot point), and the angle of rotation.

Testing:

Extends Math to include an approximate equals, Math.isClose(a,b).

Vector3d tests are found in tests/Vector3dtests.js.

tests/Vector3dtests.html runs the tests.

The first tests are testAxes, testing if axes (unit vectors (1,0,0),
(0,1,0), (0,0,1)) are perpendicular, length one, etc.

Tested vector adding (subtracting) for commutativity
(anti)-commutativity.

brb1031 committed:
Add Triple Vector tests

The vector triple product identity (Ax(BxC) + Bx(CxA) + Cx(AxB) = 0)
is tested.

The scalar triple vector product (A dot B x C) is tested for
consistency. The test is equivalent to requiring that a parallelepiped
described by 3 vectors has the same volume, no matter which edge you
call A, B, and C.

(See http://mathinsight.org/scalar_triple_product)

brb1031 committed:
Add rotation test for nonzero pivot point

Testing that rotating a point around itself should not rotate the point.

Also, test that projecting a vector onto a plane, and onto an axis
perpendicular (normal) to that plane results in vectors that satisfy:

* The axial and planar projections sum back to the original vector.

* The axial and planar projections are perpendicular.

brb1031 committed:
Add special test cases, including the zero vector

Coordinate axis vectors are added to both the unit vector list and
general vector list.  The zero vector is also added to the general list.
This is a bit of a torture test for the library, since both numerical
stability of perpendicularity and parallel(... ness?) are very
sensitive to the choice of how the lengths, angles, etc are evaluated.

The test also forces a decision as to whether the zero vector should
be perpendicular to every vector (good reason to say yes). Whether
the zero vector should be coaxial to every vector, I'm not sure, but
I picked yes.

(Vectors are coaxial if they point in either the same or exactly
opposite directions.  "Parallel" is reserved for when two coaxial
vectors point in the same direction.)

Also, replaced the test of opening angle (which is complicated,
expensive, and wrong for the zero vector) with a test for invariance of
the dot product.

brb1031 committed:
Add test for correctness of angles extracted from a vector

Methods for setting and returning vector components as spherical
coordinates are tested (and debugged!).

Testing:

Vector3d.prototype.asRThetaPhi() is designed to return the spherical
coordinates of a vector.  In order to test, Vectors are generated from
theta/phi/r, then extracted again by asRThetaPhi().

Theta is recovered (cos(theta) is compared).  Phi is recovered
(sin(theta)* phi is compared).  Both of these comparisons are more
numerically stable than comparing theta or phi directly. (For
instance, at theta = 0, north pole, nearby locations can have
radically different values of phi).

Debugging:

Corrected flaws found by testing in Vector3d.prototype.asRThetaPhi:
Eliminated unit vectoring to find theta, use the call to theta()
instead.

brb1031 commented:
Rebasing this to make it more clear.

jakebrown58 commented on vector3d.js:
```diff
@@ -1,33 +1,54 @@
+Math.TAU = Math.TAU || 2 * Math.PI;
```
I like to avoid overriding native functionality.

We need to teach you about different module patterns.  Vector is potentially a bad thing to keep prototyping so heavily if there are going to be lots of instances of it.

brb1031 replied:
Will putting it in the initial Vector3d definition cause memory/performance issues?

jakebrown58 commented on vector3d.js:
```diff
@@ -1,33 +1,54 @@
 	this.z = z;
 }
 
+Vector3d.prototype.copy = function(v){
```
We need to teach you about different code organization patterns.  Vector is potentially a bad thing to keep prototyping so heavily if there are going to be lots of instances of it.

Not something to try to tackle in this change - maybe we can do a workshop on a call soon and work that out.

jakebrown58 commented on vector3d.js:
```diff
@@ -1,33 +1,54 @@
+};
+
+Vector3d.prototype.fromRThetaPhi = function(r, theta, phi){
+	return new Vector3d(r*Math.sin(theta)*Math.cos(phi), r*Math.sin(theta)*Math.sin(phi), r*Math.cos(theta));
```
readability would be improved if you split it across lines for each param

return new Vector3d( r * Math.sin(theta) * Math.cos(phi),
     r * Math.sin(theta)*MAth.sin(phi),
    r * Math.cos(theta));

jakebrown58 commented on vector3d.js:
```diff
@@ -1,33 +1,54 @@
+	this.x = r * Math.sin(theta) * Math.cos(phi);
+	this.y = r * Math.sin(theta) * Math.sin(phi);
+	this.z = r * Math.cos(theta);
+};
```
nice and clean.  good.

jakebrown58 commented on vector3d.js:
```diff
@@ -35,9 +56,32 @@ Vector3d.prototype.sumsq = function(){
+			return Math.PI - theta;
+		}
+	}
+};
```
whitespace issues.   your tabbing here is deeper than normal.
naming - it's a function so make it an action.   getTheta would be better.   when i see 'theta', i think 'what about theta?', then i have to read the code.  if you say 'getTheta' then i assume its going to do a calc and return an angle, and i only have to read the code when i'm interested in understanding how theta is calculated.

brb1031 replied:
I branched to fix the whitespace issues in a following PR.

jakebrown58 commented on vector3d.js:
```diff
@@ -35,9 +56,32 @@ Vector3d.prototype.sumsq = function(){
+	//Azimuthal angle.
+	//Extremely cheap relative to getting both theta and phi.
+	return Math.atan2(this.y, this.x);
+};
```
getPhi

jakebrown58 commented on vector3d.js:
```diff
@@ -54,10 +98,10 @@ Vector3d.prototype.dist_squared = function(v){
 };
```
getDistSquared

jakebrown58 commented on vector3d.js:
```diff
@@ -65,113 +109,111 @@ Vector3d.prototype.dist_cubed = function(v){
 	this.y *= a;
 	this.z *= a;
+	return this;
 };
```
good on the 'Me' stuff and the return this.

jakebrown58 commented on vector3d.js:
```diff
@@ -65,113 +109,111 @@ Vector3d.prototype.dist_cubed = function(v){
+	var sintheta = Math.sqrt(1 - costheta*costheta);
+	var phi = Math.TAU * Math.random();
 	return new Vector3d(Math.cos(phi)*sintheta, Math.sin(phi)*sintheta, costheta);
 };
```
generateRandomUnitVector

generate instead of get here, because it's creating a new object versus getting something from the state of the current object.

it's awkward that we call this on the same Vector3d that we carry around state with, as I mentioned above,

jakebrown58 commented on vector3d.js:
```diff
@@ -65,113 +109,111 @@ Vector3d.prototype.dist_cubed = function(v){
-	u.scale(magnitude);
-	return u;
+	return u.scaleMe(magnitude);
 };
```
generateRandomVector

jakebrown58 commented on vector3d.js:
```diff
@@ -65,113 +109,111 @@ Vector3d.prototype.dist_cubed = function(v){
+
+Vector3d.prototype.unitMe = function(){
+	return this.scale(1 / this.magnitude());
 };
```
can magnitude be 0?   If so, you should protect magnitude from a divide by zero.

recommendation: set 1 / 0 = null, and hope the caller figured out they did something dumb.

jakebrown58 commented on vector3d.js:
```diff
@@ -65,113 +109,111 @@ Vector3d.prototype.dist_cubed = function(v){
+	}else{
+		return this.copy().scale(1 / l );
+	}
 };
```
l and 1 look the same in most fonts.   use len, otherwise 1 / l looks like 'why are we dividing 1 by 1?

jakebrown58 commented on vector3d.js:
```diff
@@ -65,113 +109,111 @@ Vector3d.prototype.dist_cubed = function(v){
+	var u2 = v2.unitVector();
+
+	return Math.acos(u1.dot(u2));
 };
```
i don't get purpose of this or why i would use it - can you explain it better?

brb1031 replied:
Opening angle is the angle between two vectors, placed tail-to-tail.  (Should be getOpeningAngle()).  Several different mathematical representations were tried; the different methods each got an "_methodDescription".  I think I've settled on one of them as the Right One.  The others are unnecessary.

jakebrown58 commented on vector3d.js:
```diff
@@ -179,12 +221,53 @@ Vector3d.prototype.cross = function(v2){
+Vector3d.prototype.increment = Vector3d.prototype.incrementMe;
+Vector3d.prototype.decrement = Vector3d.prototype.decrementMe;
+Vector3d.prototype.dist_squared = Vector3d.prototype.distSquared;
+
```
clever for now - but we should wipe out old ones subsequently in a focused PR

jakebrown58 commented on vector3d.js:
```diff
@@ -35,9 +56,32 @@ Vector3d.prototype.sumsq = function(){
-	return Math.sqrt(this.sumsq());
-}
+	return Math.sqrt(this.sumSquares());
+};
```
getMagnitude

jakebrown58 commented on vector3d.js:
```diff
@@ -35,9 +56,32 @@ Vector3d.prototype.sumsq = function(){
 };
```
getSumSquares

brb1031 committed:
Standardize Vector3d methods

Vector3d function names are standardized to conform to the following:

* <verb>Me(args) is a mutator.

* set<Sourcetype>(source) sets (mutates) an existing vector as described by source.
* get<Quantity>() calculates and returns the quantity.

* from<Coordinates>() returns a new Vector3d with coordinates in
<Coordinate> system specified by user.
* as<Targettype>() returns a new object describing the vector
as a <Targettype>.

Only the set<>() and <>Me() are mutators.  Everything else is not.

Removed some functions that were unused, and perhaps did not belong.

brb1031 commented:
Rebased to Master's head.
